man/XShm.man
.\" Copyright (c) 1991  X Consortium
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"), 
.\" to deal in the Software without restriction, including without limitation 
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense, 
.\" and/or sell copies of the Software, and to permit persons to whom the 
.\" Software furnished to do so, subject to the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL 
.\" THE X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
.\" WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF 
.\" OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
.\" SOFTWARE.
.\" 
.\" Except as contained in this notice, the name of the X Consortium shall not 
.\" be used in advertising or otherwise to promote the sale, use or other 
.\" dealing in this Software without prior written authorization from the 
.\" X Consortium.
.\"
.\" Copyright (c) 2006, Oracle and/or its affiliates. All rights reserved.
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"),
.\" to deal in the Software without restriction, including without limitation
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons to whom the
.\" Software is furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice (including the next
.\" paragraph) shall be included in all copies or substantial portions of the
.\" Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
.\" THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\" DEALINGS IN THE SOFTWARE.
.\"
.\" X Window System is a trademark of The Open Group.
.\"
.de ZN
.ie t \fB\^\\$1\^\fR\\$2
.el \fI\^\\$1\^\fP\\$2
..
.TH XShm __libmansuffix__ __xorgversion__ "X FUNCTIONS"
.SH NAME
XShmQueryExtension, XShmQueryVersion, XShmPixmapFormat, XShmAttach, XShmDetach
XShmCreateImage, XShmPutImage, XShmGetImage, XShmCreatePixmap, XShmGetEventBase - X Shared Memory extension functions
.SH SYNTAX
.nf
.LP
\&#include <X11/Xlib.h>
\&#include <sys/ipc.h>
\&#include <sys/shm.h>
\&#include <X11/extensions/XShm.h>
.LP
Bool	XShmQueryExtension(
	Display *display);
.LP
Bool XShmQueryVersion(
	Display *display;
	int *major, *minor;
	Bool *pixmaps);
.LP
Status XShmPixmapFormat(
	Display *display);
.LP
Status XShmAttach(
	Display *display;
	XShmSegmentInfo *shminfo);
.LP
Status XShmDetach(
	Display *display;
	XShmSegmentInfo *shminfo);

.LP
XImage *XShmCreateImage (
	Display *display;
	Visual *visual;
	unsigned int depth;
	int format;
	char *data;
	XShmSegmentInfo *shminfo;
	unsigned int width, height);
.LP
Status XShmPutImage(
	Display *display;
	Drawable d;
	GC gc;
	XImage *image;
	int src_x, src_y, dest_x, dest_y;
	unsigned int width, height;
	bool send_event);
.LP
Status XShmGetImage (
	Display *display;
	Drawable d;
	XImage *image;
	int x, y;
	unsigned long plane_mask);
.LP
Pixmap XShmCreatePixmap(
        Display *display;
        Drawable d;
        char *data;
	XShmSegmentInfo *shminfo;
	unsigned int width, height, depth);
.LP
Status XShmGetEventBase(
	Display *display);
.LP

.fi
.SH STRUCTURES
.nf
.ta 3i
\fIEvents:\fP
typedef struct {
    int type;               /\&* of event */
    unsigned long serial;   /\&* # of last request processed by server*/
    Bool send_event;        /\&* true if this came from a SendEvent request*/
    Display *display;       /\&* Display the event was read from */
    Drawable drawable;      /\&* drawable of request */
    int major_code;         /\&* ShmReqCode */
    int minor_code;         /\&* X_ShmPutImage */
    ShmSeg shmseg;          /\&* the ShmSeg used in the request*/
    unsigned long offset;   /\&* the offset into ShmSeg used in the request*/
} XShmCompletionEvent;
.LP
\fIa structure of type XShmSegmentInfo :\fP

typedef struct {
    ShmSeg shmseg;      /\&* resource id */
    int shmid;          /\&* kernel id */
    char *shmaddr;      /\&* address in client */
    Bool readOnly;      /\&* how the server should attach it */
} XShmSegmentInfo;

.SH DESCRIPTION
.PP
.PP
.ZN XShmQueryExtension
checks to see if the shared memory extensions are available for the
specified display.
.PP
.ZN XShmQueryVersion
returns the version numbers of the extension implementation. Shared
memory pixmaps are supported if the pixmaps argument returns true.
.PP
.ZN XShmAttach
tells the server to attach to your shared memory segment. If all goes
well, you will get a non-zero status, back and your XImage is ready
for use.
.PP
.ZN XShmDetach
tells the server to detach from your shared memory segment.
.PP
.ZN XShmPutImage
combines an image in memory with a shape of the specified drawable. If
XYBitmap format is used, the depth must be one, or a ``BadMatch''
error results. The foreground pixel in the GC defines the source for
the one bits in the image, and the background pixel defines the source
for the zero bits. For XYPixmap and ZPixmap, the depth must match the
depth of the drawable, or a ``BadMatch'' error results.
.PP
.ZN XShmGetImage
reads image data into a shared memory XImage where display is the
display of interest, drawable is the source drawable, image is the
destination XImage, x and y are offsets within the drawable, and
plane_mask defines which planes are to be read.
.PP
.ZN XShmCreateImage
allocates the memory needed for an XImage structure for the specified
display but does not allocate space for the image itself.
.PP
.ZN XShmPixmapFormat
gets the format for the server. If your application can deal with the
server pixmap data format, a shared memory segment and shminfo
structure are created.
.PP
.ZN XShmCreatePixmap
points to a pixmap which you can manipulate in all of the usual ways,
with the added bonus of being able to edit its contents directly
through the shared memory segment.
.PP
.ZN XShmGetEventBase
gets the completion event value.
.PP
.SH SEE ALSO
\fIMIT-SHM - The MIT Shared Memory Extension \fP


ChangeLog
2006-04-02  Adam Jackson  <ajax@freedesktop.org>

	* src/Xdbe.c:
	Coverity #575: Plug a leak on the error path.

	* src/XEVI.c:
	Coverity #929: Additional defensive NULL check.

2006-02-12  Alan Coopersmith  <alan.coopersmith@sun.com>

	* man/Makefile.am: 
	Bug #5628 <https://bugs.freedesktop.org/show_bug.cgi?id=5628>
	Shadow pages not created correctly when MANDIR & MANSUFFIX don't match.

2005-12-14  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for final X11R7 release candidate.

2005-12-06  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* man/Makefile.am:
	Change *man_SOURCES ==> *man_PRE to fix autotools warnings.

2005-12-03  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for X11R7 RC3 release.

2005-12-02  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	* src/Makefile.am:
	Add check and cflags for malloc(0) returning NULL.

2005-11-19  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* xext.pc.in:
	Update pkgconfig files to separate library build-time dependencies
	from application build-time dependencies.

2005-11-09  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for X11R7 RC2 release.

2005-11-01  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update pkgcheck dependencies to work with separate build roots.

2005-10-18  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for RC1 release.

2005-10-12  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* man/Makefile.am:
	Add .shadows.DONE to CLEANFILES to pass distcheck

2005-10-11  Alan Coopersmith  <alan.coopersmith@sun.com>

	* man/Makefile.am: 
	* configure.ac
	Set up cpp pre-processing of man pages
	Add shadow man pages for man pages that document multiple functions.

2005-07-08  Keith Packard  <keithp@keithp.com>

	* .cvsignore:
	* man/.cvsignore:
	* src/.cvsignore:
	* src/Makefile.am:
	Add .cvsignore files
	Switch _la_CFLAGS for AM_CFLAGS to clean up directory

2005-06-28  Alan Coopersmith  <alan.coopersmith@sun.com>

	* man/Makefile.am (man3_MANS): 
	Add man pages for DPMS Extension calls in libXext.   (Converted to
	man page format by Sun doc teams based on DPMS specification docs.)
	
2005-05-22  Alan Coopersmith  <alan.coopersmith@sun.com>

	* man/Makefile.am (man3_MANS): Change man pages to long file names

2005-01-29  Daniel Stone  <daniel@freedesktop.org>

	* XSync.c (wire_to_event):
	Sync from X.Org HEAD -- add explicit cast to prevent warning.

2004-11-22  Carl Worth  <cworth@cworth.org>

	* configure.ac: Remove AC_CONFIG_AUX_DIR (it was in the wrong
	place for automake-1.9, and it was set to the default value
	anyway).

2004-09-19  Carlos Romero  <kidcrash@freedesktop.org>

	* man/Makefile.am:
	Include man pages in EXTRA_DIST.

2004-04-13  Warren Turkal  <wt@midsouth.rr.com>

	* ImUtil.h is not in the public API and is not used in any other parts
	  of the x packages that I can see

2004-04-12  Daniel Stone  <daniel@freedesktop.org>

	* XSync.c:
	Merge from X11R6.7 - do explicit casting.

	* Bump version to 6.4.3 for xlibs 1.0.1.

2004-02-03  Jim Gettys  <jg@freedesktop.org>

	* AUTHORS needed lots of research to get the authors documented

2004-01-15  Daniel Stone  <daniel@fooishbar.org>
	* Tagging release 6.4.2 for the first freedesktop.org clientside lib
	  release.

2004-01-15  Harold L Hunt II <huntharo@msu.edu>
	* Makefile.am: Pass -no-undefined to libtool via LDFLAGS.


src/XShm.c
/* $XdotOrg: $ */
/*
 * $Xorg: XShm.c,v 1.4 2001/02/09 02:03:49 xorgcvs Exp $
 *
Copyright 1989, 1998  The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from The Open Group.
 *
 * Author:  Bob Scheifler and Keith Packard, MIT X Consortium
 */
/* $XFree86: xc/lib/Xext/XShm.c,v 1.6 2002/10/16 02:19:22 dawes Exp $ */

/* THIS IS NOT AN X CONSORTIUM STANDARD OR AN X PROJECT TEAM SPECIFICATION */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#include <stdio.h>
#include <X11/Xlibint.h>
#include <X11/ImUtil.h>
#include <X11/extensions/XShm.h>
#include <X11/extensions/shmproto.h>
#include <X11/extensions/Xext.h>
#include <X11/extensions/extutil.h>

static XExtensionInfo _shm_info_data;
static XExtensionInfo *shm_info = &_shm_info_data;
static /* const */ char *shm_extension_name = SHMNAME;

#define ShmCheckExtension(dpy,i,val) \
  XextCheckExtension (dpy, i, shm_extension_name, val)

/*****************************************************************************
 *                                                                           *
 *			   private utility routines                          *
 *                                                                           *
 *****************************************************************************/

static int close_display(Display *dpy, XExtCodes *codes);
static char *error_string(Display *dpy, int code, XExtCodes *codes,
			  char *buf, int n);
static Bool wire_to_event (Display *dpy, XEvent *re, xEvent *event);
static Status event_to_wire (Display *dpy, XEvent *re, xEvent *event);
static /* const */ XExtensionHooks shm_extension_hooks = {
    NULL,				/* create_gc */
    NULL,				/* copy_gc */
    NULL,				/* flush_gc */
    NULL,				/* free_gc */
    NULL,				/* create_font */
    NULL,				/* free_font */
    close_display,			/* close_display */
    wire_to_event,			/* wire_to_event */
    event_to_wire,			/* event_to_wire */
    NULL,				/* error */
    error_string,			/* error_string */
};

static /* const */ char *shm_error_list[] = {
    "BadShmSeg",			/* BadShmSeg */
};

static XEXT_GENERATE_FIND_DISPLAY (find_display, shm_info, shm_extension_name, 
				   &shm_extension_hooks, ShmNumberEvents, NULL)

static XEXT_GENERATE_CLOSE_DISPLAY (close_display, shm_info)

static XEXT_GENERATE_ERROR_STRING (error_string, shm_extension_name,
				   ShmNumberErrors, shm_error_list)


static Bool
wire_to_event (Display *dpy, XEvent *re, xEvent *event)
{
    XExtDisplayInfo *info = find_display (dpy);
    XShmCompletionEvent	*se;
    xShmCompletionEvent	*sevent;

    ShmCheckExtension (dpy, info, False);

    switch ((event->u.u.type & 0x7f) - info->codes->first_event) {
    case ShmCompletion:
	se = (XShmCompletionEvent *) re;
	sevent = (xShmCompletionEvent *) event;
	se->type = sevent->type & 0x7f;
	se->serial = _XSetLastRequestRead(dpy,(xGenericReply *) event);
	se->send_event = (sevent->type & 0x80) != 0;
	se->display = dpy;
	se->drawable = sevent->drawable;
	se->major_code = sevent->majorEvent;
	se->minor_code = sevent->minorEvent;
	se->shmseg = sevent->shmseg;
	se->offset = sevent->offset;
    	return True;
    }
    return False;
}

static Status
event_to_wire (Display *dpy, XEvent *re, xEvent *event)
{
    XExtDisplayInfo *info = find_display (dpy);
    XShmCompletionEvent	*se;
    xShmCompletionEvent	*sevent;

    ShmCheckExtension (dpy, info, 0);

    switch ((re->type & 0x7f) - info->codes->first_event) {
    case ShmCompletion:
    	se = (XShmCompletionEvent *) re;
	sevent = (xShmCompletionEvent *) event;
    	sevent->type = se->type | (se->send_event ? 0x80 : 0);
    	sevent->sequenceNumber = se->serial & 0xffff;
    	sevent->drawable = se->drawable;
    	sevent->majorEvent = se->major_code;
    	sevent->minorEvent = se->minor_code;
    	sevent->shmseg = se->shmseg;
    	sevent->offset = se->offset;
    	return True;
    }
    return False;
}

/*****************************************************************************
 *                                                                           *
 *		    public Shared Memory Extension routines                  *
 *                                                                           *
 *****************************************************************************/

Bool XShmQueryExtension (Display *dpy /*  int *event_basep, *error_basep */)
{
    XExtDisplayInfo *info = find_display (dpy);

    if (XextHasExtension(info)) {
/*	*event_basep = info->codes->first_event;
	*error_basep = info->codes->error_event; */
	return True;
    } else {
	return False;
    }
}


int XShmGetEventBase(Display *dpy)
{
    XExtDisplayInfo *info = find_display (dpy);

    if (XextHasExtension(info)) {
	return info->codes->first_event;
    } else {
	return -1;
    }
}


Bool XShmQueryVersion(
    Display *dpy,
    int *majorVersion,
    int *minorVersion,
    Bool *sharedPixmaps)
{
    XExtDisplayInfo *info = find_display (dpy);
    xShmQueryVersionReply rep;
    register xShmQueryVersionReq *req;

    ShmCheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(ShmQueryVersion, req);
    req->reqType = info->codes->major_opcode;
    req->shmReqType = X_ShmQueryVersion;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return False;
    }
    *majorVersion = rep.majorVersion;
    *minorVersion = rep.minorVersion;
    *sharedPixmaps = rep.sharedPixmaps ? True : False;
    UnlockDisplay(dpy);
    SyncHandle();
    return True;
}


int XShmPixmapFormat(Display *dpy)
{
    XExtDisplayInfo *info = find_display (dpy);
    xShmQueryVersionReply rep;
    register xShmQueryVersionReq *req;

    ShmCheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(ShmQueryVersion, req);
    req->reqType = info->codes->major_opcode;
    req->shmReqType = X_ShmQueryVersion;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return 0;
    }
    UnlockDisplay(dpy);
    SyncHandle();
    if (rep.sharedPixmaps &&
	(rep.majorVersion > 1 || rep.minorVersion > 0))
	return rep.pixmapFormat;
    return 0;
}


Status XShmAttach(Display *dpy, XShmSegmentInfo *shminfo)
{
    XExtDisplayInfo *info = find_display (dpy);
    register xShmAttachReq *req;

    ShmCheckExtension (dpy, info, 0);

    LockDisplay(dpy);
    GetReq(ShmAttach, req);
    req->reqType = info->codes->major_opcode;
    req->shmReqType = X_ShmAttach;
    req->shmseg = shminfo->shmseg = XAllocID(dpy);
    req->shmid = shminfo->shmid;
    req->readOnly = shminfo->readOnly ? xTrue : xFalse;
    UnlockDisplay(dpy);
    SyncHandle();
    return 1;
}


Status XShmDetach(Display *dpy, XShmSegmentInfo *shminfo)
{
    XExtDisplayInfo *info = find_display (dpy);
    register xShmDetachReq *req;

    ShmCheckExtension (dpy, info, 0);

    LockDisplay(dpy);
    GetReq(ShmDetach, req);
    req->reqType = info->codes->major_opcode;
    req->shmReqType = X_ShmDetach;
    req->shmseg = shminfo->shmseg;
    UnlockDisplay(dpy);
    SyncHandle();
    return 1;
}

static int _XShmDestroyImage (XImage *ximage)
{
	Xfree((char *)ximage);
	return 1;
}

#define ROUNDUP(nbytes, pad) ((((nbytes) + ((pad) - 1)) / (pad)) * (pad))

XImage *XShmCreateImage (
    register Display *dpy,
    register Visual *visual,
    unsigned int depth,
    int format,
    char *data,
    XShmSegmentInfo *shminfo,
    unsigned int width,
    unsigned int height)
{
    register XImage *image;

    image = (XImage *)Xcalloc(1, (unsigned)sizeof(XImage));
    if (!image)
	return image;
    image->data = data;
    image->obdata = (char *)shminfo;
    image->width = width;
    image->height = height;
    image->depth = depth;
    image->format = format;
    image->byte_order = dpy->byte_order;
    image->bitmap_unit = dpy->bitmap_unit;
    image->bitmap_bit_order = dpy->bitmap_bit_order;
    image->bitmap_pad = _XGetScanlinePad(dpy, depth);
    image->xoffset = 0;
    if (visual) {
	image->red_mask = visual->red_mask;
	image->green_mask = visual->green_mask;
	image->blue_mask = visual->blue_mask;
    } else {
	image->red_mask = image->green_mask = image->blue_mask = 0;
    }
    if (format == ZPixmap)
	image->bits_per_pixel = _XGetBitsPerPixel(dpy, (int)depth);
    else
	image->bits_per_pixel = 1;
    image->bytes_per_line = ROUNDUP((image->bits_per_pixel * width),
				    image->bitmap_pad) >> 3;
    _XInitImageFuncPtrs(image);
    image->f.destroy_image = _XShmDestroyImage;
    return image;
}

Status XShmPutImage (
    register Display *dpy,
    Drawable d,
    GC gc,
    register XImage *image,
    int src_x, int src_y, int dst_x, int dst_y,
    unsigned int src_width, unsigned int src_height,
    Bool send_event)
{
    XExtDisplayInfo *info = find_display (dpy);
    XShmSegmentInfo *shminfo = (XShmSegmentInfo *)image->obdata;
    register xShmPutImageReq *req;

    ShmCheckExtension (dpy, info, 0);
    if (!shminfo) return 0;

    LockDisplay(dpy);
    FlushGC(dpy, gc);
    GetReq(ShmPutImage, req);
    req->reqType = info->codes->major_opcode;
    req->shmReqType = X_ShmPutImage;
    req->drawable = d;
    req->gc = gc->gid;
    req->srcX = src_x;
    req->srcY = src_y;
    req->srcWidth = src_width;
    req->srcHeight = src_height;
    req->dstX = dst_x;
    req->dstY = dst_y;
    req->totalWidth = image->width;
    req->totalHeight = image->height;
    req->depth = image->depth;
    req->format = image->format;
    req->sendEvent = send_event;
    req->shmseg = shminfo->shmseg;
    req->offset = image->data - shminfo->shmaddr;
    UnlockDisplay(dpy);
    SyncHandle();
    return 1;
}


Status XShmGetImage(
    register Display *dpy,
    Drawable d,
    XImage *image,
    int x, int y,
    unsigned long plane_mask)
{
    XExtDisplayInfo *info = find_display (dpy);
    XShmSegmentInfo *shminfo = (XShmSegmentInfo *)image->obdata;
    register xShmGetImageReq *req;
    xShmGetImageReply rep;
    register Visual *visual;

    ShmCheckExtension (dpy, info, 0);
    if (!shminfo) return 0;

    LockDisplay(dpy);
    GetReq(ShmGetImage, req);
    req->reqType = info->codes->major_opcode;
    req->shmReqType = X_ShmGetImage;
    req->drawable = d;
    req->x = x;
    req->y = y;
    req->width = image->width;
    req->height = image->height;
    req->planeMask = plane_mask;
    req->format = image->format;
    req->shmseg = shminfo->shmseg;
    req->offset = image->data - shminfo->shmaddr;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return 0;
    }
    visual = _XVIDtoVisual(dpy, rep.visual);
    if (visual) {
    	image->red_mask = visual->red_mask;
    	image->green_mask = visual->green_mask;
    	image->blue_mask = visual->blue_mask;
    } else {
	image->red_mask = image->green_mask = image->blue_mask = 0;
    }
    UnlockDisplay(dpy);
    SyncHandle();
    return 1;
}

Pixmap XShmCreatePixmap (
    register Display *dpy,
    Drawable d,
    char *data,
    XShmSegmentInfo *shminfo,
    unsigned int width, unsigned int height, unsigned int depth)
{
    XExtDisplayInfo *info = find_display (dpy);
    Pixmap pid;
    register xShmCreatePixmapReq *req;

    ShmCheckExtension (dpy, info, 0);

    LockDisplay(dpy);
    GetReq(ShmCreatePixmap, req);
    req->reqType = info->codes->major_opcode;
    req->shmReqType = X_ShmCreatePixmap;
    req->drawable = d;
    req->width = width;
    req->height = height;
    req->depth = depth;
    req->shmseg = shminfo->shmseg;
    req->offset = data - shminfo->shmaddr;
    pid = req->pid = XAllocID(dpy);
    UnlockDisplay(dpy);
    SyncHandle();
    return pid;
}
